### PR TITLE
Add PDB to gateway chart

### LINKED ISSUE #44469: PodDisruptionBudget support in gateway helm chart
(This is used to request new product features, please visit <https://discuss.istio.io> for questions on using Istio)

**Describe the feature request**
Current gateway [helm chart](https://github.com/istio/istio/tree/master/manifests/charts/gateway) doesn't support `PodDisruptionBudget`, 

**Describe alternatives you've considered**
We need to manually deploy `PodDisruptionBudget`

**Affected product area (please put an X in all that apply)**

[ ] Ambient
[ ] Docs
[X] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Affected features (please put an X in all that apply)**

[ ] Multi Cluster
[ ] Virtual Machine
[ ] Multi Control Plane

**Additional context**

### PR DESCRIPTION
**Pull request by hanxiaop**
**Please provide a description of this PR:**
Fix https://github.com/istio/istio/issues/44469

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [ ] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [x] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

**Please check any characteristics that apply to this pull request.**

- [ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

**Comment by zirain** [comment]
IMHO, -1 for bring this into chart:
1. cannot put everything into chart, should focus on the core features, once bring in a new resource type, you need support customization for every filed
2. it's easy to manage things like this with modern gitops tools like TF/fluxcd

**Comment by hanxiaop** [comment]
@zirain I don't have a strong preference regarding the addition of PDB to the gateway chart, as it's not very necessary. I was thinking the resource should be important to gateways and can provide a better installation experience if users have specific needs to it.

> it's easy to manage things like this with modern gitops tools like fluxcd

For this, I think users who prefer to manage resources using tools can simply disable the PDB option in the chart and deploy their customized PDB as needed. If I'm not using such a tool, I would have to manually deploy the PDB myself. Alternatively, we could consider disabling the PDB by default to allow for easier customization.

I'm not actually very familiar with the balance between the out-of-the-box functionalities and customization of the charts, and if this is not a great practice I can close it.

**Comment by infa-ddeore** [comment]
PDB is kind of default that almost every tool provided with its helm chart, 

check [nginx](https://github.com/kubernetes/ingress-nginx/blob/main/charts/ingress-nginx/templates/) or [ambassador](https://github.com/datawire/ambassador-chart/blob/master/templates/pdb.yaml), they do provide PDB

**Comment by deva26** [comment]
+1 for this support in gateway charts. Most of the production use cases, it makes absolute sense.

It sounds reasonable to expect parity when users switch from istioctl manifest generate to helm charts

**Comment by hanxiaop** [comment]
Reopen for more discussions

**Comment by kfaseela** [comment]
> IMHO, -1 for bring this into chart:
> 
> 1. cannot put everything into chart, should focus on the core features, once bring in a new resource type, you need support customization for every filed
> 2. it's easy to manage things like this with modern gitops tools like fluxcd

@zirain  looks this is a parity between helm and istioctl install? One path already supports configuring PDB?

**Review by jacob-delgado — APPROVED** [review]

File: manifests/charts/gateway/templates/poddisruptionbudget.yaml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+{{- if .Values.defaultPodDisruptionBudget.enabled }}
+{{- if (semverCompare ">=1.21-0" .Capabilities.KubeVersion.GitVersion) }}
+apiVersion: policy/v1
```
Comment (howardjohn) [inline manifests/charts/gateway/templates/poddisruptionbudget.yaml:3]: I would just only support v1 at this point. Its new feature (so no backwards compat needed) and support for 1.20 is going away in this release anyways

File: manifests/charts/gateway/values.yaml [hunk 2]
```diff
@@ -95,3 +95,6 @@ networkGateway: ""
 imagePullPolicy: ""
 
 imagePullSecrets: []
+
+defaultPodDisruptionBudget:
```
Comment (howardjohn) [inline manifests/charts/gateway/values.yaml:99]: Why is it **default**PodDisruptionBudget? Either there is a PDB or there isn't, not sure what "default" means here?

Comment (hanxiaop) [reply manifests/charts/gateway/values.yaml]: I copied this from the `gateways` chart, and it appears that it was originally "podDisruptionBudget" but eventually changed to "defaultDisruptionBudget." Regardless, I have revised it to PDB for better clarity. Thanks.

**Comment by costinm** [comment]
I also believe users who want to customize PDB should do it using a
separate chart or yaml, where they have full control and
can adjust it without reinstalling the gateway.

But since moving forward I think the K8S Gateway with Istiod managed
install is going to be the best option - and this chart
and istioctl will fade away - I don't mind if it's added. But if we add it,
it should not be just 'enabled' - but allow user
to specify a min, max and unhealthyPodEvictionPolicy - we can keep default
values if needed. I don't know why
users would set this in helm values instead of a proper yaml - but at least
will be the same keywords and syntax as the real
object.

**Comment by therealmitchconnors** [comment]
I think it should be possible to install production-grade istio directly from helm without the need to modify or add resources.  This PR is a great step in that direction, and we should consider adding PDBs to istiod managed gateways as well.

**Comment by jacob-delgado** [comment]
While we can't support every feature within K8s within our helm charts, we should maintain parity with anything created from istioctl for helm. PDBs are pretty important for gateways as are nodeSelector, tolerations, affinity and anti-affinity.

**Comment by hanxiaop** [comment]
> I don't mind if it's added. But if we add it, it should not be just 'enabled' - but allow user to specify a min, max and unhealthyPodEvictionPolicy - we can keep default values if needed.

@costinm Thanks for the suggestion. I have revised based on the comment. Please take another look.

**Comment by costinm** [comment]
We keep optimizing for install and demo - and not for operations and day 2.

And in the process - we actually make install worse for users who want to
operate Istio: they need
to learn our newly invented API for configuring pod disruption ( expressed
as a Istio-specific
helm values.yaml API with minimal testing and validation, or as an even
more Istio specific operator yaml )
instead of using the APIs that K8S provide - and have validations and much
better documentation.

I use helm to install every application I can, so I did a list for PDBs in
my cluster, the only ones I found
are created by Istio plus one created by konectivity-agent in kube-system.

While I have some experience installing and consider myself advanced user -
I don't know how I could
turn this off, since values.yaml has a default value (1). I'm sure there is
an obscure helm 'set' command
to do it - my point is most users who would need a PDB will have a much
easier time having an
'operations' helm chart or separate yaml files where they can use well
documented and tested
K8S APIS with PDB, HPA, Certificates, etc - instead of the equivalent and
mostly untested
istio-invented APIs in values.yaml.

In any case - I'm not blocking this PR, just want people to think about it
and start considering values.yaml as
an API that requires the same amount of testing and documentation and
stability as any other API, as well
as the same criteria on 'is it worth inventing our own API for something
that already exists upstream' ?

**Comment by costinm** [comment]
BTW - I'm less concerned about this PR because I believe best practice will
be to use the K8S Gateway resource to handle gateway creation,
and that path is doing only the minimal thing required ( Deployment,
Service, ServiceAccount - both optional and merged with
user-specified CRs). Not to mention it auto-manages the ports on the
service and has well-defined and standard status.

I would be strongly against adding PDBs, HPAs - and a large set of
annotations - to replicate what operator and istio
helm does for the K8S API.

**Review by costinm — APPROVED** [review]
Maybe add some comments - including what is the proper value ( is 1 supposed to be modified ?), that user can put the entire fragment, how they can turn it off and use PDB yaml directly, etc.

I know the rest of values.yaml 'API' is undocumented, so not blocking on this.

**Review by howardjohn — CHANGES_REQUESTED** [review]
Can we put it off by default?

Its actually broken to have `replicas=1, minAvailable=1`. Kubernetes nodes will stop scaling down. We have done it in the past but no need to repeat our mistakes

**Comment by psbrar99** [comment]
@hanxiaop : I hope this will be backported to release-1.17 and release-1.16 branches. Can you confirm?

**Comment by hanxiaop** [comment]
@psbrar99 I think we can cherry-pick this since this fixes the parity with different install methods. I'm going to combine this one and its follow-up, and do a manual cherry-pick for the release branch.

**Comment by barth12** [comment]
Hi guys, great job! Should it be in this chart already: https://artifacthub.io/packages/helm/istio-official/gateway?

**Pull request merged**